pkg/controller/core.oam.dev/v1alpha2/appdeployment/appdeployment_controller.go

```diff
@@ -0,0 +1,561 @@
+			if err := r.updateStatus(ctx, appDeployment); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		if err := r.deleteRevisions(ctx, appDeployment, diff.Del); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.applyRevisions(ctx, appDeployment, diff.Mod); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.applyRevisions(ctx, appDeployment, diff.Add); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	appDeployment.Status.Phase = oamcore.PhaseCompleted
+	appDeployment.Status.Placement = makePlacement(
+		append(append(diff.Add, diff.Mod...), diff.Unchanged...),
+	)
+
+	if appDeployment.Spec.Traffic != nil {
+		if err := r.applyTraffic(ctx, appDeployment); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, r.updateStatus(ctx, appDeployment)
+}
+
+func (r *Reconciler) handleFinalizer(ctx context.Context, appd *oamcore.AppDeployment) error {
+	if !slice.ContainsString(appd.Finalizers, appDeploymentFinalizer, nil) {
+		return nil
+	}
+	// our finalizer is present, so lets handle any external dependency
+	if err := r.deleteExternalResources(ctx, appd); err != nil {
+		// if fail to delete the external dependency here, return with error
+		// so that it can be retried
+		return err
+	}
+
+	appd.ObjectMeta.Finalizers = removeString(appd.ObjectMeta.Finalizers, appDeploymentFinalizer)
+	return errors.Wrap(r.Client.Update(ctx, appd), errUpdateFinalizer)
+}
+
+func (r *Reconciler) deleteExternalResources(ctx context.Context, appd *oamcore.AppDeployment) error {
+	var revsDel []*revision
+	for _, p := range appd.Status.Placement {
+		for _, c := range p.Clusters {
+			if isHostCluster(c.ClusterName) {
+				continue
+			}
+			revsDel = append(revsDel, newRevision(p.RevisionName, c.ClusterName, c.Replicas))
+		}
+	}
+
+	return r.deleteRevisions(ctx, appd, revsDel)
+}
+
+func (r *Reconciler) getClientForCluster(ctx context.Context, cluster, ns string) (client.Client, error) {
+	c, err := r.getCluster(ctx, cluster, ns)
+	if err != nil {
+		return nil, err
+	}
+
+	key := client.ObjectKey{
+		Name:      c.Spec.KubeconfigSecretRef.Name,
+		Namespace: ns,
+	}
+	secret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, key, secret); err != nil {
+		return nil, err
+	}
+	return clustermanager.GetClient(secret.Data["config"])
+}
+
+func (r *Reconciler) deleteRevisions(ctx context.Context, appd *oamcore.AppDeployment, revisions []*revision) (err error) {
+	for _, rev := range revisions {
+		klog.InfoS("delete revision", "revision", rev.RevisionName, "cluster", rev.ClusterName)
+
+		workloads, err := r.getWorkloadsForRevision(ctx, rev.RevisionName, appd.Namespace)
+		if err != nil {
+			return err
+		}
+
+		var kubecli client.Client
+		if isHostCluster(rev.ClusterName) {
+			kubecli = r.Client
+		} else {
+			kubecli, err = r.getClientForCluster(ctx, rev.ClusterName, appd.Namespace)
+			if err != nil {
+				return err
+			}
+		}
+
+		for _, wl := range workloads {
```

Comment:
There's a case, that the workload instance is resued in the new revision.  If so , it will be deleted here while it should still be in use.

 I think we need to check for it.

Reply:
I can't think of any case like that. Can you give a more concrete example?